Bottom: 7b36af729c37138677f219ec4b7327cca11eb2b1
Top:    052c67a095bde4900c8fbf20aea8301a747e95ce
Author: Ben Greear <greearb@candelatech.com>
Date:   2014-03-20 16:18:32 -0700

ath10k: add helper method to grab debug stats.

It can be nice to update the firmware's stats while
debugging other bits of the driver, so add helper method
to do this.

Signed-off-by: Ben Greear <greearb@candelatech.com>


---

diff --git a/drivers/net/wireless/ath/ath10k/debug.c b/drivers/net/wireless/ath/ath10k/debug.c
index e251155..12461be 100644
--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -413,7 +413,7 @@ free:
 
 static int ath10k_debug_fw_stats_request(struct ath10k *ar)
 {
-	unsigned long timeout, time_left;
+	unsigned long timeout;
 	int ret;
 
 	lockdep_assert_held(&ar->conf_mutex);
@@ -426,19 +426,9 @@ static int ath10k_debug_fw_stats_request(struct ath10k *ar)
 		if (time_after(jiffies, timeout))
 			return -ETIMEDOUT;
 
-		reinit_completion(&ar->debug.fw_stats_complete);
-
-		ret = ath10k_wmi_request_stats(ar, ar->fw_stats_req_mask);
-		if (ret) {
-			ath10k_warn(ar, "could not request stats (%d)\n", ret);
+		ret = ath10k_refresh_peer_stats(ar);
+		if (ret)
 			return ret;
-		}
-
-		time_left =
-		wait_for_completion_timeout(&ar->debug.fw_stats_complete,
-					    1 * HZ);
-		if (!time_left)
-			return -ETIMEDOUT;
 
 		spin_lock_bh(&ar->data_lock);
 		if (ar->debug.fw_stats_done) {
@@ -502,6 +492,27 @@ static int ath10k_fw_stats_release(struct inode *inode, struct file *file)
 	return 0;
 }
 
+int ath10k_refresh_peer_stats(struct ath10k *ar)
+{
+	int ret;
+	unsigned long time_left;
+
+	reinit_completion(&ar->debug.fw_stats_complete);
+	ret = ath10k_wmi_request_stats(ar, ar->fw_stats_req_mask);
+	if (ret) {
+		ath10k_warn(ar, "could not request stats (%d)\n", ret);
+		return ret;
+	}
+
+	/* ret means 'time-left' here */
+	time_left =
+		wait_for_completion_timeout(&ar->debug.fw_stats_complete, 1*HZ);
+	if (time_left == 0)
+		return -ETIMEDOUT;
+
+	return 0;
+}
+
 static ssize_t ath10k_fw_stats_read(struct file *file, char __user *user_buf,
 				    size_t count, loff_t *ppos)
 {
diff --git a/drivers/net/wireless/ath/ath10k/debug.h b/drivers/net/wireless/ath/ath10k/debug.h
index 75c89e3..f74c159 100644
--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -186,4 +186,7 @@ static inline void ath10k_dbg_dump(struct ath10k *ar,
 {
 }
 #endif /* CONFIG_ATH10K_DEBUG */
+
+int ath10k_refresh_peer_stats(struct ath10k *ar);
+
 #endif /* _DEBUG_H_ */
